### PR TITLE
fix: 8385 added button variant support in table column type button

### LINKED ISSUE #8385: [Bug]: Table Button column type does not have Button variant option
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Table Button column type does not have button variant option

### Steps To Reproduce

```markdown
1.Launch appsmith app in the release/Prod environment 
2.Drag and drop table widget 
3.Change any of the column type to button 
4.Check for the button variant option and observe
```

### Environment

_No response_

### Version

_No response_

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> Added button variant support for column type button

Fixes #8385 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Added test cases.
- Tested locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/8385-button-variant-for-table-columtype-button 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.75 **(0.03)** | 37.2 **(0.04)** | 35.46 **(0.02)** | 56.01 **(0.03)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableStyledWrappers.tsx | 89.53 **(2.03)** | 56.2 **(0)** | 83.64 **(2.94)** | 89.53 **(2.03)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableUtilities.tsx | 37.28 **(-0.82)** | 23.48 **(-0.41)** | 16.22 **(0)** | 35.48 **(-0.88)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/index.tsx | 30.07 **(0)** | 9.48 **(0.05)** | 30.19 **(0)** | 29.77 **(0)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/propertyConfig.ts | 95.83 **(-0.09)** | 100 **(0)** | 93.33 **(-0.22)** | 95.83 **(-0.09)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/propertyUtils.ts | 43.92 **(15.45)** | 25.47 **(14.69)** | 37.5 **(10.83)** | 43.17 **(15.83)**</details>

### CODE REVIEW

**Review by Tooluloope — APPROVED** [review]

**Comment by shwetha-ramesh** [comment]
@vibhandikyash Tested this PR. Button variant option is now visible but there are 2 issues that needs to be looked at.

1. By default for the button widget the font that appears on the widget is BOLD. Where as in table widget the font of the _label_ of the button does not appear bold. However this is noticed in Release as well. @somangshu @sbalaji1192 can give a heads up if this needs to be implemented.
image

2. Not able to change the _Label_ color. When verified in release, able to change the color where as in DP the color of the _label_ remains the same even after different selection is made. 
https://loom.com/share/8b70fb6448d640c4ae5947814d5e423b

**Moving to Inprogress.**

**Comment by vibhandikyash** [comment]
> @vibhandikyash Tested this PR. Button variant option is now visible but there are 2 issues that needs to be looked at.
> 
> 1. By default for the button widget the font that appears on the widget is BOLD. Where as in table widget the font of the _label_ of the button does not appear bold. However this is noticed in Release as well. @somangshu @sbalaji1192 can give a heads up if this needs to be implemented.
>    image
> 2. Not able to change the _Label_ color. When verified in release, able to change the color where as in DP the color of the _label_ remains the same even after different selection is made.
>    https://loom.com/share/8b70fb6448d640c4ae5947814d5e423b
> 
> **Moving to Inprogress.**

1: Font weight is different in table (400) and button text (500)

2. In table, button color and text color is given by old method, now we are using button color and variant in button, menu button, icon button, there is no option for text color

@somangshu any comments?

**Comment by somangshu** [comment]
@vibhandikyash @shwetha-ramesh 
- For point 1: This is as is from before. @momcilo-appsmith @UXFalcon please let us know if anychanges are required here
- For point 2: We need consistency between the properties, In this PR we were introducing the button variant property and hence other such properties in favour of consistency should exist. Let me know if there is confusion in understanding, Ill check it out in detail

**Comment by momcilo-appsmith** [comment]
@somangshu The buttons should be the same, no matter where they are, in the table or on the canvas.

**Comment by somangshu** [comment]
> @somangshu The buttons should be the same, no matter where they are, in the table or on the canvas.

@vibhandikyash @shwetha-ramesh please check for consistency, We could take it up separately if you see many differences

**Comment by vibhandikyash** [comment]
> > @somangshu The buttons should be the same, no matter where they are, in the table or on the canvas.
> 
> @vibhandikyash @shwetha-ramesh please check for consistency, We could take it up separately if you see many differences

@somangshu okay, currently I've updated the font weight for consistency. We shall take other design stuff later on.

**Review by Tooluloope — APPROVED** [review]

**Comment by chandannkumar** [comment]
Tested the DP and observed that Shwetha's callouts has not been fixed yet and issues remains same as observed in previous testing
@vibhandikyash @Tooluloope

**Comment by Tooluloope** [comment]
@momcilo-appsmith @vibhandikyash is this how the height of the Button column type is meant to be?.

@chandannkumar The No.1 Comment is handled Since it's handled in Button Color

@vibhandikyash Let's remove the Label color from The Property pane for Button Column type

There should be padding at the top and bottom

**Deploy Preview**
Screenshot 2022-02-17 at 08 40 42


**Release**
Screenshot 2022-02-17 at 08 47 49



cc: @somangshu

**Comment by somangshu** [comment]
@Tooluloope @chandannkumar @shwetha-ramesh we need the button and its properties to remain consistent with the the actual widget, When you guys are talking the decision or testing the feature, make sure the comparison happens. even in terms of the UX. Ex if the label color is removed, the button in the table widget needs to act similar to the button widget

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash @Tooluloope Test this PR and please find the status below,

* Original issue - Button variant is now available for the button column type 
* Issues from Shwetha :  
        Point 1 - Font size difference - **Fixed** 
        Point 2 - Remove label  colour - **Not fixed** 
        Point 3 - There should be Padding for button column type - **Not Fixed**

**New Migration issue:** 

Point 4 - Pre-request - Old app should have table widget with icon , menu columns , Three Text widget with values (PRIMARY, SECONDARY and TERTIARY)

Open any existing app which has columns Icon, Menu button type and which is binded with text widget - Icon to bind with {{Text23.Text}} and Menu column type is binded with {{Text2.text}}, open the old app in the DP and open table property pane and try to change the button variant of icon with {{Text1.text}} and observe icon gets added to the menu button column type and choosing the icon to none the icon will not clear and it will remain

https://www.loom.com/share/dfa95b9ece0346e3bb154ec65a46c7b3

**Review by Tooluloope — DISMISSED** [review]

**Comment by somangshu** [comment]
@vibhandikyash what is the update here, How long approximately we are going to take to close this ?

**Comment by Tooluloope** [comment]
@vibhandikyash Could you help resolve @YogeshJayaseelan Comments?

**Comment by vibhandikyash** [comment]
Yes, I'll look into it..

**Comment by techbhavin** [comment]
@somangshu @Tooluloope  @YogeshJayaseelan  

Point 2: Fixed
Point 3 : Fixed 

Point 4:   (**this is not a migration issue** )In Column Type menuButton and IconButton has used the same property `iconName` in `app/client/src/widgets/TableWidget/widget/propertyConfig.ts`. 

so that’s why it will show an icon in menuButton. if we remove defaultIcoName from propertyConfig then IconButton will not show default `add` icon.

**Comment by Tooluloope** [comment]
> @somangshu @Tooluloope @YogeshJayaseelan
> 
> Point 2: Fixed Point 3 : Fixed
> 
> Point 4: (**this is not a migration issue** )In Column Type menuButton and IconButton has used the same property `iconName` in `app/client/src/widgets/TableWidget/widget/propertyConfig.ts`.
> 
> so that’s why it will show an icon in menuButton. if we remove defaultIcoName from propertyConfig then IconButton will not show default `add` icon.

Thanks for pointing this out @techbhavin

**Review by Tooluloope — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this PR and please find the status below,

Original issue - The button variant is now available for the button column type
Issues from Shwetha :
Point 1 - Font size difference - *Fixed*
Point 2 - Remove label colour - *Fixed*
Point 3 - There should be Padding for button column type - *Fixed*

@somangshu @vibhandikyash @Tooluloope @shwetha-ramesh  New observation: In terms of consistency the table button column type does not have the following properties - Button colour, Border radius, Shadow colour, Icon and icon alignments

**Comment by somangshu** [comment]
@YogeshJayaseelan please detail out the scenario, what is not available for what column type? as below

- Text type
    - missing properties
         - text color

It will be a better context to understand

**Comment by somangshu** [comment]
@techbhavin @Tooluloope @YogeshJayaseelan for the icon issue pointed out above, Here are my thought

I checked the DP, Feels like the behaviour is Appropriate
- in the same column the icon is persistent, that should be expected for the dev.
- in another column, the icon is at a default and the dev needs to choose the icon again. again the behaviour is predictable

Other observation,
- Did we remove the default icon from here, The initial selection which is black might not be the way to proceed, i see this conflict with what we have on release and prod, @sbalaji1192 need some inputs here, Can you review.

@techbhavin please point out the code you were showing me earlier.

**Comment by dilippitchika** [comment]
@somangshu you are right, we should have a default icon. I don't think we should remove it.

**Comment by techbhavin** [comment]
@somangshu  please refer [Here](https://github.com/appsmithorg/appsmith/blob/release/app/client/src/widgets/TableWidget/widget/propertyConfig.ts#L820)

**Comment by YogeshJayaseelan** [comment]
@techbhavin Please find the properties fields which are missing for the button column type 

Styles:
        * Border Radius
        * Box Shadow
        * Shadow Colour
        * Icon 
        * Placement 
        * Icon Alignment

**Comment by sbalaji1192** [comment]
@somangshu I have discussed with @techbhavin on how we can change iconName based on the columnType dynamically. he will be making that change.
@YogeshJayaseelan Let's create an issue on the Table widget v2 epic to add these properties to button columnType. We don't have to do it in Table v1 as part of this issue CC @dilippitchika

File: app/client/src/widgets/TableWidget/widget/propertyUtils.ts [hunk 1]
```diff
@@ -206,6 +207,38 @@ export const updateColumnStyles = (
...
+    propertyValue,
+  );
+  if (updateDerivedColumnsHookArr) {
+    propertiesToUpdate = [
+      ...updateDerivedColumnsHookArr,
+      ...propertiesToUpdate,
+    ];
+  }
+
+  if (!iconName && propertyValue === "iconButton") {
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/widget/propertyUtils.ts:232]: We need to reset when user switches to `menuButton`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_PropertyPane_spec.js [hunk 2]
```diff
@@ -348,7 +348,81 @@ describe("Table Widget property pane feature validation", function() {
     cy.get(publish.backToEditor).click();
   });
 
-  it("14. Table-Delete Verification", function() {
+  it("14. Verify table columy type button with button variant", function() {
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_PropertyPane_spec.js:351]: typo in column

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js [hunk 3]
```diff
@@ -104,30 +104,11 @@ describe("Table Widget property pane feature validation", function() {
...
-      .click({ force: true })
-      .clear()
-      // following wait is required to reproduce #9526
-      .wait(600)
-      .type(color2);
-    cy.get(widgetsPage.tableBtn).should("have.css", "color", color2);
-  });
-
-  it("4. Table widget triggeredRow property should be accessible", function() {
+  it("3. Table widget triggeredRow property should be accessible", function() {
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js:107]: Why are we removing the buttonLabelColor?

File: app/client/src/widgets/TableWidget/widget/index.tsx [hunk 4]
```diff
@@ -169,6 +169,8 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
                 this.onCommandClick(rowIndex, action, onComplete),
               backgroundColor: cellProperties.buttonColor || "transparent",
               buttonLabelColor: cellProperties.buttonLabelColor || "#FFFFFF",
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/widget/index.tsx:171]: we removed this property on this commit `buttonLabelColor`

File: app/client/src/widgets/TableWidget/widget/index.tsx [hunk 5]
```diff
@@ -223,7 +225,7 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
               borderRadius: cellProperties.borderRadius,
               boxShadow: cellProperties.boxShadow,
               boxShadowColor: cellProperties.boxShadowColor,
-              iconName: cellProperties.iconName || "add",
+              iconName: cellProperties.iconName || undefined,
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/widget/index.tsx:228]: can we use empty string?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_PropertyPane_spec.js [hunk 6]
```diff
@@ -348,7 +348,81 @@ describe("Table Widget property pane feature validation", function() {
...
+    // Open property pane
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_PropertyPane_spec.js:352]: Need to add a case to verify that updateIconNameHook is doing its job.

File: app/client/src/widgets/TableWidget/widget/propertyUtils.ts [hunk 7]
```diff
@@ -206,8 +207,44 @@ export const updateColumnStyles = (
   }
   return;
 };
+
+// Select default Icon Name if column type is Icon Button
+export function updateIconNameHook(
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/widget/propertyUtils.ts:212]: Please add a jest case for this hook.

**Review by sbalaji1192 — COMMENTED** [review]
Please improve the PR title.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js (hunk 3 above)
Comment (techbhavin) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js:107]: `buttonLabelColor` was used in old table properties. 

As per QA comment in current PR : 
We need consistency between the properties and in the latest button, there is no option like buttonLabelColor. it is replaced by a button variant

File: app/client/src/widgets/TableWidget/widget/index.tsx (hunk 5 above)
Comment (techbhavin) [reply app/client/src/widgets/TableWidget/widget/index.tsx:227]: @sbalaji1192 
we cannot use an empty string, if we do that button component will show empty space mentioned in the below image 

image

File: app/client/src/widgets/TableWidget/widget/index.tsx [hunk 8]
```diff
@@ -168,7 +168,8 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
               onCommandClick: (action: string, onComplete: () => void) =>
                 this.onCommandClick(rowIndex, action, onComplete),
               backgroundColor: cellProperties.buttonColor || "transparent",
-              buttonLabelColor: cellProperties.buttonLabelColor || "#FFFFFF",
+              buttonVariant:
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/widget/index.tsx:171]: This will break the applications that were using the buttonLabelColor, right?

Comment (techbhavin) [reply app/client/src/widgets/TableWidget/widget/index.tsx:171]: No, that will not break anything. we removed that property and the current button design works on button variants so if there is data in DSL, it's useless.

**Review by sbalaji1192 — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@sbalaji1192 @techbhavin can you share me the latest deployment build here

**Comment by techbhavin** [comment]
@YogeshJayaseelan  you can  use this

https://appsmith-97aeeuz2c-get-appsmith.vercel.app/

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this PR and please find the status below,

Original issue - The button variant is now available for the button column type

Point 1 - Font size difference - **Fixed**
Point 2 - Remove label colour - **Fixed**
Point 3 - There should be Padding for button column type - **Fixed**
Point 4 - By default icon is displayed for Icon column type - **Fixed**
Point 5 - Missing button properties will be handled in Table V2

**Pull request merged**